src/librustc_codegen_llvm/back/write.rs

```diff
@@ -399,6 +399,9 @@ impl CodegenContext {
 }
 
 struct DiagnosticHandlers<'a> {
+    #[allow(dead_code)]
+    // This value is not actually dead, llcx has pointers to it and needs these pointers to be alive
+    // until Drop is executed on this object
     inner: Box<(&'a CodegenContext, &'a Handler)>,
```

Comment:
Do you mean the heap allocation itself? This is really confusing, maybe a raw pointer should be kept instead, with a manual `Drop` impl that uses `Box::from_raw`.

Reply:
Also maybe the comment should be phrased to be about the callback having a copy of the pointer or something. It's not really "self-referential" insofar as much as it is scoping a resource for the duration C++ needs it.